react/components/ModalContext.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { createContext, useReducer, useContext } from 'react'
+
+interface State {
+  open: boolean
+}
+
+const DEFAULT_STATE = {
+  open: false,
+}
+
+function mergeStates(state?: Partial<State>) {
+  return {
+    ...DEFAULT_STATE,
+    ...state,
+  }
+}
+
+interface OpenAction {
+  type: 'OPEN_MODAL'
+}
+
+interface CloseAction {
+  type: 'CLOSE_MODAL'
+}
+
+type Action = OpenAction | CloseAction
+type Dispatch = (action: Action) => void
+
+const ModalStateContext = createContext<State>(DEFAULT_STATE)
+const ModalDispatchContext = createContext<Dispatch | undefined>(undefined)
+
+function modalContextReducer(state: State = DEFAULT_STATE, action: Action) {
+  switch (action.type) {
+    case 'OPEN_MODAL': {
+
+      return {
+        ...state,
+        open: true,
+      }
+    }
+
+    case 'CLOSE_MODAL':
+        return {
+          ...state,
+          open: false,
+        }
+
+    default:
+      return state
+  }
+}
+
+interface ModalContextProviderProps {
+  initialState?: Partial<State>
+}
+
+export const ModalContextProvider: React.FC<ModalContextProviderProps> = props => {
+  const { children, initialState } = props
+  const [state, dispatch] = useReducer(modalContextReducer, mergeStates(initialState))
+
+  return (
+    <ModalStateContext.Provider value={state}>
+      <ModalDispatchContext.Provider value={dispatch}>
+        {children}
+      </ModalDispatchContext.Provider>
+    </ModalStateContext.Provider>
+  )
+}
+
+export function useModalDispatch() {
+  const context = useContext(ModalDispatchContext)
+  // if (typeof context === 'undefined') {
+  //   throw Error('useModalDispatch must be used within a ModalContextProvider')
+  // }
```

Comment:
Should this be a comment yet? Its a good idea to think about whether this verification should be here or not because if this is not here in the `modal-layout`, the verification needs to happen on components that use it.
I do think it is a good idea to have it here in this case

Reply:
This is good but is not working for some reason that I don't know why The component receives the context  as `undefined` sometimes